### PR TITLE
Implement QWLayerSurfaceV1/QWLayerShellV1

### PR DESCRIPTION
**Pull request by wineee**
wlroots is using `namespace` as a variable name, which is inlegal in C++
https://gitlab.freedesktop.org/wlroots/wlroots/-/issues/3651

### CODE REVIEW

File: src/CMakeLists.txt [hunk 1]
```diff
@@ -101,6 +101,13 @@ set(PRIVATE_HEADERS
     types/qwinputdevice_p.h
 )
 
+pkg_search_module(WLR_PROTOCOLS wlr-protocols)
+if (WLR_PROTOCOLS_FOUND)
+    ws_generate(server wlr-protocols unstable/wlr-layer-shell-unstable-v1.xml wlr-layer-shell-unstable-v1-protocol)
+    list(APPEND HEADERS types/qwlayershellv1.h)
```
Comment (zccrs) [inline src/CMakeLists.txt:107]: We can add a new value for that protocols files. for an example:

````
set(PROTOCOL_HEADERS
         ....
)

list(APPEND HEADERS ${PROTOCOL_HEADERS})
````

**Review by zccrs — CHANGES_REQUESTED** [review]

File: src/types/qwlayershellv1.cpp [hunk 2]
```diff
@@ -0,0 +1,238 @@
...
+}
+
+QWLayerSurfaceV1 *QWLayerSurfaceV1::get(wlr_layer_surface_v1 *handle)
+{
+    return QWLayerSurfaceV1Private::map.value(handle);
+}
+
+QWLayerSurfaceV1 *QWLayerSurfaceV1::from(wlr_layer_surface_v1 *handle)
+{
+    if (handle == nullptr)
```
Comment (zccrs) [inline src/types/qwlayershellv1.cpp:188]: We don't needs check the handle value in there, it's value should check from caller.

File: src/types/qwlayershellv1.cpp [hunk 3]
```diff
@@ -0,0 +1,238 @@
...
+        return nullptr;
+    if (auto o = get(handle))
+        return o;
+    return new QWLayerSurfaceV1(handle, false);
+}
+
+QWLayerSurfaceV1 *QWLayerSurfaceV1::from(wl_resource *resource)
+{
+    return from(wlr_layer_surface_v1_from_resource(resource));
```
Comment (zccrs) [inline src/types/qwlayershellv1.cpp:197]: If the return value of wlr_layer_surface_v1_from_resource maybe is nullptr, we can check the return value in there, isn't check it in `QWLayerSurfaceV1::from(wlr_layer_surface_v1 *handle)`

**Review by zccrs — CHANGES_REQUESTED** [review]

File: src/types/qwlayershellv1.cpp (hunk 3 above)
Comment (wineee) [reply src/types/qwlayershellv1.cpp]: `wlr_layer_surface_v1_from_resource`  will  `aborts`  if the resource doesn't have the correct type 
`wlr_layer_surface_v1_try_from_wlr_surface`  maybe  `returns NULL`

**Review by zccrs — APPROVED** [review]

**Pull request merged**